### PR TITLE
Joshdellaz/feature/leop more robust

### PR DESCRIPTION
**Pull request by joshdellaz**
Changes made to ensure reliable LEOP as per TX2-MD-028 [CC] Ver. 3.00 Ex-Alta 2 Mission Operations Plan.

Readability improvements as well.

### CODE REVIEW

File: ex2_system/source/leop/leop.c [hunk 1]
```diff
@@ -37,64 +38,74 @@
...
-        if ((switchstatus(sw) != 1) && (getStatus_retries != MAX_RETRIES)) {
-            ex2_log("Check #%d: %c not deployed\n", &getStatus_retries, sw);
-            ex2_log("Activated %c\n", sw);
-            activate(sw);
-            vTaskDelay(TWENTY_SEC_DELAY);
-        } else if ((switchstatus(sw) != 1) && (getStatus_retries == MAX_RETRIES)) {
-            ex2_log("Check #%d: %c not deployed, exiting the LEOP sequence.\n", &getStatus_retries, sw);
-            return false;
-        }
+        ex2_log("Activated burnwire %c\n", sw);
```
Comment (Taylorr82) [inline ex2_system/source/leop/leop.c:47]: Why did you remove the part where the code checks the deployment status?

File: ex2_system/source/leop/leop.c [hunk 2]
```diff
@@ -37,64 +38,74 @@
...
-                return false;
-            }
+/**
+ * @brief
+ *      Log all switch statuses after LEOP
+ * @details
+ *      Logs all switch statuses after LEOP
+ * @return void
+ */
+void log_switch_status() {
```
Comment (Taylorr82) [inline ex2_system/source/leop/leop.c:78]: I don't understand the purpose of this change

File: ex2_system/source/leop/leop.c [hunk 3]
```diff
@@ -107,20 +118,16 @@ bool deploy_all_deployables() {
...
-            // If all hard switch have been deployed, set eeprom flag to TRUE
-            eeprom_set_leop_status();
-            return true;
-        }
-        return false;
-    }
-    //    else if (eeprom_get_leop_status() == true) {
-    else if (eeprom_flag == true) {
+        // If leop sequence was never executed, doit
+        deploy_all_deployables();
```
Comment (Taylorr82) [inline ex2_system/source/leop/leop.c:125]: Why did you remove all the error handling? If LEOP doesn't work, we don't want to indicate that it did

**Review by Taylorr82 — CHANGES_REQUESTED** [review]

File: ex2_system/source/leop/leop.c (hunk 1 above)
Comment (yunqingyi) [reply ex2_system/source/leop/leop.c:47]: mhm, we should include it here. ex2_log has been changed to sys_log since then as well

**Review by yunqingyi — CHANGES_REQUESTED** [review]
We should be doing error handling and logging sequentially so we understand which step LEOP failed at in the code. I noticed that burnwire currents were removed, I think leaving it there will give us more insight into the potential reason of why LEOP failed.

File: ex2_system/source/leop/leop_eeprom.c [hunk 4]
```diff
@@ -24,16 +24,15 @@ bool eeprom_get_leop_status() {
...
         }
-    }
-    else {
+    } else {
         return false;
     }
 }
 
-bool eeprom_set_leop_status() {
+bool eeprom_set_leop_status(bool status) {
```
[the remark this answers is not in the document]
Comment (joshdellaz) [reply ex2_system/source/leop/leop_eeprom.c:32]: Yup. We concluded that discussion after I clocked out for the evening. Will make that change

File: ex2_system/source/leop/leop.c (hunk 1 above)
Comment (joshdellaz) [reply ex2_system/source/leop/leop.c:47]: What if the switch gives a faulty reading, or the deployable moves by 1mm to de-trigger the switch? Then boom, no antenna deployment and our satellite is dead to us.

**Comment by joshdellaz** [comment]
If leop fails, we're screwed, and there's nothing we can do in retrospect. What will current even tell us? We know the voltage is 5V and the resistor is 10 ohms and we will have tested the exact same hardware on the bench. I don't see justification, so I am simplifying.

File: ex2_system/source/leop/leop.c (hunk 3 above)
Comment (joshdellaz) [reply ex2_system/source/leop/leop.c:125]: The final switch statuses are all we really care about, no? There should be no reason for leop to not complete. Whether or not it was successful, we will know by the switch statuses

File: ex2_system/source/leop/leop.c (hunk 1 above)
Comment (yunqingyi) [reply ex2_system/source/leop/leop.c:47]: original design was to start over again if we get stuck deploying one switch (since they needed to be done in sequence so different deployables don't get in each other's way). I'm okay with simplifying it to 5 attempts and moving on if it's not successful.

File: ex2_system/source/leop/leop.c (hunk 3 above)
Comment (yunqingyi) [reply ex2_system/source/leop/leop.c:125]: I think the error handling can stay, we can remove the return false statements though so LEOP can proceed onto the next switch. We should keep the logs in error handling to understand what went wrong and when.

**Review by yunqingyi — CHANGES_REQUESTED** [review]
We can remove the return false statements so LEOP can proceed onto the next switch even if one fails. 
I think we should keep the current readings - if the switch status is faulty but the current readings are within expected range, then we know there is a chance the deployment was successful.

File: ex2_system/source/leop/leop.c (hunk 3 above)
Comment (joshdellaz) [reply ex2_system/source/leop/leop.c:125]: What's the purpose of error handling if it'll always return true?

I don't understand what utility it provides us to have logs in between deployment attempts. We just care about the final state after LEOP is done. What would we do with the other information from in between attempts? How would that inform our actions post-leop?

**Comment by joshdellaz** [comment]
The current is not helpful in determining deployment state because
- the current will _always_ be the same, unless the resistors fail. A resistor would only fail if we put it through hell before launch, which we will not be doing.
- testing has shown that current has never varied during any deploy() action
- deployment success rate is heavily dependent on length of burnwire activation and the position of the burnwire on the resistor

File: ex2_system/source/leop/leop.c (hunk 3 above)
Comment (yunqingyi) [reply ex2_system/source/leop/leop.c:125]: The same reason for including so much information in our beacon. Generally speaking it's better to have more information. It's also a good practice to know how many attempts were made and how many failures we saw in the process, this is a standard typically followed in the industry, and maybe Robert can speak to this more but I think it should be a standard we follow in our code. I prefer to have a complete sequence of events, even if the final outcome is the same.

**Comment by Taylorr82** [comment]
> The current is not helpful in determining deployment state because
> 
>     * the current will _always_ be the same, unless the resistors fail. A resistor would only fail if we put it through hell before launch, which we will not be doing.
> 
>     * testing has shown that current has never varied during any deploy() action
> 
>     * deployment success rate is heavily dependent on length of burnwire activation and the position of the burnwire on the resistor

Wouldn't it be good to know a resistor failed? Determine the root cause for a failure. Whether it's a switch failed to inform us failure, resistor failure, maybe even a burnwire slipped off the resistor failure. My point is, error reporting is valuable. Maybe it won't help us, maybe it will, maybe it's useful for ex-alta 3.

File: ex2_hal/athena/equipment_handler/include/deployablescontrol.h [hunk 5]
```diff
@@ -22,7 +22,7 @@
 
 #include "HL_gio.h"
 
-#define DEPLOYABLE_BURNWIRE_DELAY_MS 6500 // Delay 6.5 seconds according to systems requirements
+#define DEPLOYABLE_BURNWIRE_DELAY_MS 30000 // Delay 30 seconds according to prior hardware testing
```
Comment (Taylorr82) [inline ex2_hal/athena/equipment_handler/include/deployablescontrol.h:25]: Do we have enough power to do the maximum number of attempts at 30 seconds?

File: main/main.c [hunk 6]
```diff
@@ -111,11 +111,8 @@ void ex2_init(void *pvParameters) {
     init_csp();
 
     /* LEOP */
-
-#if EXECUTE_LEOP == 1
-    if (execute_leop() != true) {
-        // TODO: Do what if leop fails?
-    }
+#ifdef EXECUTE_LEOP == 1
```
Comment (Taylorr82) [inline main/main.c:114]: incorrect use of ifdef
